lib/collection/src/tests/mod.rs

```diff
@@ -342,3 +351,68 @@ fn check_version_upgrade() {
         &"0.4.2".parse().unwrap()
     ));
 }
+
+#[tokio::test(flavor = "multi_thread")]
+#[rstest]
+#[case::auto_optimization_threads(None, ShardStatus::Yellow)]
+#[case::no_optimization_threads(Some(0), ShardStatus::Green)]
+#[case::single_optimization_thread(Some(1), ShardStatus::Yellow)]
+async fn test_shard_status_based_on_optimization(
+    #[case] max_optimization_threads_value: Option<usize>,
+    #[case] expected_shard_status: ShardStatus,
+) {
+    let collection_name = "test_collection".to_string();
+    let collection_dir = Builder::new().prefix(&collection_name).tempdir().unwrap();
+
+    let mut config = create_collection_config();
+    config.optimizer_config.max_optimization_threads = max_optimization_threads_value;
+    config.optimizer_config.default_segment_number = 4;
+
+    let current_runtime: Handle = Handle::current();
+
+    let payload_index_schema_dir = Builder::new().prefix("qdrant-test").tempdir().unwrap();
+    let payload_index_schema_file = payload_index_schema_dir.path().join("payload-schema.json");
+    let payload_index_schema =
+        Arc::new(SaveOnDisk::load_or_init_default(payload_index_schema_file).unwrap());
+
+    let collection_config = Arc::new(TokioRwLock::new(config.clone()));
+
+    let shard = LocalShard::build(
+        0,
+        collection_name.clone(),
+        collection_dir.path(),
+        collection_config.clone(),
+        Arc::new(Default::default()),
+        payload_index_schema.clone(),
+        current_runtime.clone(),
+        current_runtime.clone(),
+        ResourceBudget::default(),
+        config.optimizer_config.clone(),
+    )
+    .await
+    .unwrap();
+
+    let upsert_ops = upsert_operation();
+    let hw_acc = HwMeasurementAcc::new();
+    shard
+        .update(upsert_ops.into(), true, hw_acc.clone())
+        .await
+        .unwrap();
+
+    // This should trigger segment merge optimizer if we have optimizer threads (i.e. != Some(0))
+    {
+        let mut collection_config_guard = collection_config.write().await;
+        collection_config_guard
+            .optimizer_config
+            .default_segment_number = 1;
+    }
```

Comment:
This by itself will not trigger the optimizer, because it doesn't fire an update.

We should probably do the operation shown [here](https://qdrant.tech/documentation/concepts/collections/#grey-collection-status) as well to actually _trigger_.

Reply:
> optimizer, because it doesn't fire an update

Yes but I'm doing this afterwards and tests are passing:
```rs
// This recreates optimizers based on updated collection config
shard.on_optimizer_config_update().await.unwrap();
```

Reply:
> I wanted to drop the guard automatically instead of calling `drop()` myself so I kept that part inside a block. Do you know any better syntax for this?

Yes, just this:

```rust
    // Change configuration and trigger merge optimizer
    collection_config.write().await.optimizer_config.default_segment_number = 1;
    let info = shard.local_shard_info().await;
    assert_eq!(info.status, ShardStatus::Green);
    shard.on_optimizer_config_update().await.unwrap();
```